File: Piko/Sources/SFOutputNode.m
//
//  SFOutputNode.m
//
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

#import "SFOutputNode.h"

#import "SFInput.h"
#import "SFOutput.h"

@implementation SFOutputNode

- (instancetype)init
{
    SFInput *input = [[SFInput alloc] initWithName:@"input"];
    SFOutput *output = [[SFOutput alloc] initWithName:@"output"];
    self = [super initWithInputs:@[input] outputs:@[output]];
    if (self != nil) {
        _input = input;
        _output = output;
    }
    return self;
}

- (BOOL)rendersConstantFrame
{
    return NO;
}

- (BOOL)renderForPlayer:(SFPlaybackManager *)player
                  index:(UInt64)index
                  count:(NSInteger)count
                     to:(double *)buffer
{
    SFNode *inputNode = _input.source.owner;
    if (inputNode != nil) {
        return [inputNode renderForPlayer:player
                                    index:index
                                    count:count
                                       to:buffer];
    }

    for (NSInteger i = 0; i < count; i++) {
        buffer[i] = 0.0;
    }
    return YES;
}

@end

File: Piko/Sources/SFOutputNode.swift
//
//  SFOutputNode.swift
//
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

import Foundation

@objc class SFOutputNode: SFNode {

    let input: SFInput
    let output: SFOutput

    init() {
        input = SFInput(name: "input")
        output = SFOutput(name: "output")
        super.init(inputs: [input], outputs: [output])
    }

    override func renderForPlayer(player: SFPlaybackManager,
                                   index: UInt64,
                                   count: Int,
                               to buffer: UnsafeMutablePointer<Double>) -> Bool {
        if let node = input.source?.owner! {
            return player.render(node, index: index, count: count, to: buffer)
        } else {
            for i in 0..<count {
                buffer[i] = 0.0
            }
            return true
        }
    }
}
